[SYCL] Move variadic call diagnostics to delayed diagnostics, fixes.

erichkeane opened the pull request:
One of our downstreams requires using the variadic call diagnostics from
SYCL, however the SemaSYCL AST walker ends up being messy. This patch
initially removes the check from the AST walker and instead implements
its in normal SEMA via the delayed diagnostics mechanism.

However, while evaluating this, we discovered that there are quite a few
issues with how the diagnostics are emitted, so this patch includes
quite a few cleanups to make sure the diagnostics are emitted properly.
This results in some additional notes in some of the test.

Additionally, the asm checks were previously implemented in BOTH delayed
diagnostics as well as the AST walker, since the bugs in each aligned to
have full coverage.  This patch ends up removing it from the AST walking
code since the other bugs were fixed.

Signed-off-by: Erich Keane <erich.keane@intel.com>

premanandrao commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -372,18 +372,6 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
-    return true;
-  }
-
   // The call graph for this translation unit.
```
Were these dead code and safe to remove?

erichkeane replied:
They were not, but are now :)  When This is the case in the commit message where we had done 2 versions of diagnostics for the inline assembly, but bugs in each made them not cover all cases, but combined, they seemed to.  My fix to the delayed diagnostics made the assembly error handling in normal 'Sema' cover the cases that this caught.

premanandrao commented on clang/test/SemaSYCL/sycl-cconv.cpp:
```diff
@@ -8,7 +8,7 @@ __inline __cdecl __attribute__((sycl_device)) int foo() { return 0; }
 
 void bar() {
-  printf("hello\n"); // expected-no-error
+  //printf("hello\n"); // expected-no-error
```
Did you mean to comment  out that call?

erichkeane replied:
Woops, nope!  Shouldn't have changed that, thanks for the catch!

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1420,18 +1407,20 @@ Sema::DeviceDiagBuilder Sema::SYCLDiagIfDeviceCode(SourceLocation Loc,
          "Should only be called during SYCL compilation");
-  DeviceDiagBuilder::Kind DiagKind = [this] {
-    if (isKnownEmitted(*this, dyn_cast<FunctionDecl>(CurContext)))
+  FunctionDecl *FD = dyn_cast<FunctionDecl>(CurContext);
```
Shouldn't it be CurLexicalContext?

erichkeane replied:
I believe it SHOULD be getCurLexicalContext, but the current implementation was CurContext.  I couldn't come up with a reproducer that I was able to justify changing the status quo here.

Fznamznon replied:
I should double check, because it was before long vacation, but in my patch I've implemented a new usage of deferred diagnostics - diagnosing of `__float128` usage in device code (https://github.com/intel/llvm/pull/971/files#diff-643c0b5e3365e57856a6871266a81b35) and it didn't work before I changed it to `getCurLexicalContext` .

erichkeane replied:
Neat!  Thanks for providing justification.   I'll upload an updated patch momentarily.

erichkeane commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -14027,6 +14027,11 @@ Sema::BuildCallToObjectOfClassType(Scope *S, Expr *Obj,
   // If this is a variadic call, handle args passed through "...".
   if (Proto->isVariadic()) {
+    // Diagnose variadic calls in SYCL.
+    if (getLangOpts().SYCLIsDevice && !isUnevaluatedContext())
```
Well, there isn't really a 'decl' to check!  This is the operator call for a type, so the decl doesn't actually necessarily exist (and we don't actually have it).  Since we aren't marking call operators a 'known good' there is no reason to check it.

Fznamznon replied:
Oh, okay then

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1404,13 +1391,13 @@ static bool isKnownEmitted(Sema &S, FunctionDecl *FD) {
   if (!FD)
     return true; // Seen in LIT testing
 
-  if (FD->hasAttr<SYCLDeviceAttr>() || FD->hasAttr<SYCLKernelAttr>())
```
Because otherwise we end up double diagnosing.  The template part has to be first, otherwise SCYLDevice marked templates will diagnose both when going through stage 1 and then when being instantiated.

Fznamznon replied:
Am I right that the second diagnosing happens in SemaSYCL when we marked all device functions with `SYCLDevice` attribute?

erichkeane replied:
No, only the emission of the diagnostic happens when functions are marked with the device attribute.  The repro I had was something like (from memory, pardon the typos):

    <template>
    __attribute__((sycldevice)) void some_func() {
        variadic_call();
   }

In this case, during phase 1 (when the template is being parsed, but is uninstantiated), this would be diagnosed  immediately thanks to the marked line being above the template check.

THEN, it would be called (instantiated), and would diagnose AGAIN there.  THEN, in cases where it was the actual kernel, we would diagnose a THIRD time (since it was the "OpenCLKernel"!).  

Not that my changes to return K_Nop in the function below might make the template check here not necessary, but I wasn't positive so I left it.

Fznamznon replied:
Ok, thanks for the explanation.

Fznamznon commented:
I actually thought that we also will fix this `FIXME` https://github.com/intel/llvm/blob/8803f62b8f8004a51ed2fc6365085bb7cec015a8/clang/lib/Sema/SemaType.cpp#L4688 with this change.

erichkeane commented:
> 
> 
> I actually thought that we also will fix this `FIXME`
> 
> https://github.com/intel/llvm/blob/8803f62b8f8004a51ed2fc6365085bb7cec015a8/clang/lib/Sema/SemaType.cpp#L4688
> with this change.

I don't think so, do we?  That is OpenCL diagnostics, which I didn't change.  In fact, I'd be surprised if the OpenCL path for these diagnostic repairs don't have similar issues to what I fixed.

Fznamznon commented:
> > I actually thought that we also will fix this `FIXME`
> > https://github.com/intel/llvm/blob/8803f62b8f8004a51ed2fc6365085bb7cec015a8/clang/lib/Sema/SemaType.cpp#L4688
> > 
> > with this change.
> 
> I don't think so, do we? That is OpenCL diagnostics, which I didn't change. In fact, I'd be surprised if the OpenCL path for these diagnostic repairs don't have similar issues to what I fixed.

@bader , do you have any thoughts?

bader commented:
Variadic functions are not allowed in OpenCL and SYCL device code.
OpenCL uses exception and allows `printf`.
There is no such exception for SYCL in the spec. It says that `printf` functionality is covered by `stream` class.
According to my interpretation of the SYCL spec, `printf` is no different from any other variadic function and should be diagnosed in the SYCL device code.
From what I see, today compiler allows using `printf` in the device code w/o reporting any errors, which seems to be a bug mentioned in the `FIXME` comment.

> That is OpenCL diagnostics, which I didn't change. In fact, I'd be surprised if the OpenCL path for these diagnostic repairs don't have similar issues to what I fixed.
@erichkeane, I'm not sure I understand: what "issues" do you refer here?

erichkeane commented:
> 
> 
> Variadic functions are not allowed in OpenCL and SYCL device code.
> OpenCL uses exception and allows `printf`.
> There is no such exception for SYCL in the spec. It says that `printf` functionality is covered by `stream` class.
> According to my interpretation of the SYCL spec, `printf` is no different from any other variadic function and should be diagnosed in the SYCL device code.
> From what I see, today compiler allows using `printf` in the device code w/o reporting any errors, which seems to be a bug mentioned in the `FIXME` comment.
> 
> > That is OpenCL diagnostics, which I didn't change. In fact, I'd be surprised if the OpenCL path for these diagnostic repairs don't have similar issues to what I fixed.
> > @erichkeane, I'm not sure I understand: what "issues" do you refer here?

The ones I fixed in this patch :), particularly when it comes to templates.  OpenCL uses a similar (though not identical) delayed diagnostics mechanism that I have not evaluated (since improving OpenCL is not my goal here) but look similar enough to the issues that the SYCL code has currently (at least until this patch is commited).

I believe that FIXME is NOT fixed, since it is diagnosing OpenCL (which this patch does not), and I believe that fixing it out of the scope of both this patch and this repo.

bader commented:
> OpenCL uses a similar (though not identical) delayed diagnostics

Really? OpenCL compiler work with "device code" only, so the reason to "delay" diagnostics might be quite different from the SYCL device compiler, if any. Could you point the source code where OpenCL delays the diagnostics, please?

> I believe that FIXME is NOT fixed,

Right.

> since it is diagnosing OpenCL (which this patch does not), and I believe that fixing it out of the scope of both this patch and this repo.

The FIXME comment is not about fixing OpenCL behavior, but rather fixing SYCL compiler behavior, which I described in my previous comment: https://github.com/intel/llvm/pull/998#issuecomment-574211449. Having said that I think this patch is the right place to fix diagnostic for `printf`, which is "variadic call". Does it make sense?

erichkeane commented:
> 
> 
> > OpenCL uses a similar (though not identical) delayed diagnostics
> 
> Really? OpenCL compiler work with "device code" only, so the reason to "delay" diagnostics might be quite different from the SYCL device compiler, if any. Could you point the source code where OpenCL delays the diagnostics, please?

Ah, I was thinking of CUDA.  I see a bunch of delay-diagnostics code for CUDA.  Don't know why I remembered 'OpenCL' :).

> > I believe that FIXME is NOT fixed,
> 
> Right.
> 
> > since it is diagnosing OpenCL (which this patch does not), and I believe that fixing it out of the scope of both this patch and this repo.
> 
> The FIXME comment is not about fixing OpenCL behavior, but rather fixing SYCL compiler behavior, which I described in my previous comment: [#998 (comment)](https://github.com/intel/llvm/pull/998#issuecomment-574211449). Having said that I think this patch is the right place to fix diagnostic for `printf`, which is "variadic call". Does it make sense?

I guess I still don't get what you're trying to say.  It seems that we explicitly exclude ourselves from that check, and my patch does NOT exclude printf, so it should be doing what you want, right?  

As a side note, that OpenCL check doesn't work correctly in C++... it seems to exclude ANY function named printf, not just the one in stdio.h :)

erichkeane commented:
> 
> 
> > I guess I still don't get what you're trying to say. It seems that we explicitly exclude ourselves from that check, and my patch does NOT exclude printf, so it should be doing what you want, right?
> 
> It was excluded by initial implementation. Previous implementation just skipped the diagnostics to allow using `printf` on in the host part of the code (i.e. before we had any diagnostics for SYCL device code). I think what Marya proposes is to include the fix for this bug into this PR.
> 
> We either should:
> 
>     * just remove the comment as deferred diagnositcs for variadic function calls was implemented already. Could you make sure that we have a LIT tests covering calling `printf` from the device code?

Got it.  I'll remove the comment.  sycl-cconv.cpp already tests with printf.

> 
>     * add separate deferred check for the `printf` calls. Instead of skipping function type check in SYCL device mode, add a new if for SYCL and use deferred diagnostics engine to report the error at this place.
> 
> 
> BTW, ideally we should make use of existing diagnostics implemented for CUDA instead of adding new one: https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/DiagnosticSemaKinds.td#L7631. Can we make this CUDA-specific diagnostic generic and re-use for SYCL? I think we can re-use the code reporting this diagnostic.

I can definitely switch to that diagnostic if you wish, but it isn't saying the same thing or diagnosing the same thing. That is diagnosing a function DECLARATION, not a function CALL.  Thus, it is a different error diagnosis.  Also note that I'm not adding a error message here.

> > As a side note, that OpenCL check doesn't work correctly in C++... it seems to exclude ANY function named printf, not just the one in stdio.h :)
> 
> Feel free to report this. Tagging @AnastasiaStulova.
@AnastasiaStulova  : That likely needs to make sure that the printf is externC and in the global namespace.

Fznamznon commented:
> I can definitely switch to that diagnostic if you wish, but it isn't saying the same thing or diagnosing the same thing. That is diagnosing a function DECLARATION, not a function CALL. Thus, it is a different error diagnosis. Also note that I'm not adding a error message here.

Yes, I'm not sure that we can reuse it because we should diagnose calls, not declarations. As far as I understand in CUDA in most cases device code is marked explicitly, meanwhile in SYCL any function can magically become a device function if called from a kernel.

Overall, maybe I like place for deferred diagnostic where OpenCL does diagnosing a bit more because IMO it simplifies searching through SEMA code, but if it works and it's already done it's ok too.

AnastasiaStulova commented:
> As a side note, that OpenCL check doesn't work correctly in C++... it seems to exclude ANY function named printf, not just the one in stdio.h

We don't have standard includes in OpenCL and also there are numerous ways to declare BIFs so the way to check is string compare the name. In the C++ mode it should indeed have C linkage.

erichkeane commented:
Clang doesn't typically check where a function is defined/declared, so the string check is fine, just not specific enough. You need to make sure it is C linkage, not in a namespace, not a member function, etc. Note it could likely be a friend declaration (should likely be allowed), though I wonder about a friend definition...

AnastasiaStulova commented:
Yes indeed this still needs to be done!

> You need to make sure it is C linkage, not in a namespace, not a member function, etc. Note it could likely be a friend declaration (should likely be allowed), though I wonder about a friend definition...

Would some of these not be redundant? I would expect any C linkage functions wouldn't be a member function? In other words C linkage functions won't be parsed with C++ features i.e. class members, templates, etc...

erichkeane commented:
> 
> 
> Yes indeed this still needs to be done!
> 
> > You need to make sure it is C linkage, not in a namespace, not a member function, etc. Note it could likely be a friend declaration (should likely be allowed), though I wonder about a friend definition...
> 
> Would some of these not be redundant? I would expect any C linkage functions wouldn't be a member function? In other words C linkage functions won't be parsed with C++ features i.e. class members, templates, etc...

Yes, some are.  "C" linkage can't be applied to templates or class members, but still would apply to friend functions I would expect.  Also, inside a namespace a function can still be extern "C".

AnastasiaStulova commented:
>  Also, inside a namespace a function can still be extern "C".

I am not sure that should be diagnosed. At least for the namespaces it still doesn't seem to work as regular C++ functions would. There will only be one C linkage function with a certain name regardless of the namespace. Although if we want to be on the safe side we might add this to the check.

erichkeane commented:
> > Also, inside a namespace a function can still be extern "C".
> 
> I am not sure that should be diagnosed. At least for the namespaces it still doesn't seem to work as regular C++ functions would. There will only be one C linkage function with a certain name regardless of the namespace. Although if we want to be on the safe side we might add this to the check.

I'm not an OpenCL language designer, so take my opinion as just that :)  You're likely right that "extern C" is sufficient, as long as that is what you mean.  Friend definitions can get strange I think, so just make sure that the decision is made having played with them.  In reality, extern "C" is likely sufficient.
